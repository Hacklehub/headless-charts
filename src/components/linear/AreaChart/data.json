[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
