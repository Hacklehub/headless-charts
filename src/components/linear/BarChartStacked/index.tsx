import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { max, sum } from 'd3-array';
import { scaleBand, scaleLinear } from 'd3-scale';
import { select, selectAll } from 'd3-selection';
import { useCallback, useEffect } from 'react';
import useTooltip, { TooltipObjectType } from '../../../hooks/useTooltip';

import { format } from 'd3-format';
import { stack } from 'd3';
import { transition } from 'd3-transition';

interface DataItem {
  [key: string]: any;
}

interface AxisItems {
  key: string;
  className?: string;
  axis?: string;
  axisTicks?: number;
}

interface Drawing {
  duration?: number;
  delay?: number;
}

interface DataLabel {
  text?: (data: DataItem, column: AxisItems) => string;
}

interface ReferenceLines {
  x?: number;
  className?: string;
}

interface StackedArrayItem {
  0: number;
  1: number;
  data: DataItem;
}

interface StackedDataItem {
  start: number;
  end: number;
  data: DataItem;
  index: number;
}

interface BarChartStackedProps {
  data: DataItem[];
  id: string;
  className?: string;
  direction?: string;
  padding?: {
    left: number;
    right: number;
    bottom: number;
    top: number;
    bar: number;
  };
  margin?: {
    left: number;
    right: number;
    top: number;
    bottom: number;
  };
  referenceLines?: ReferenceLines[];
  waterfall?: boolean;
  x: AxisItems[];
  tickFormat?: string;
  y: AxisItems;
  tooltip?: TooltipObjectType;
  drawing?: Drawing;
  dataLabel?: DataLabel;
}

const BarChartStacked = ({
  data = [],
  id,
  className,
  direction = 'right',
  padding = {
    left: 0,
    right: 0,
    bottom: 0,
    top: 0,
    bar: 0.3,
  },
  margin = {
    left: 60,
    right: 20,
    top: 20,
    bottom: 40,
  },
  referenceLines = [],
  waterfall,
  x,
  tickFormat,
  y,
  tooltip,
  drawing = undefined,
  dataLabel,
}: BarChartStackedProps) => {
  const { onMouseOver, onMouseMove, onMouseLeave } = useTooltip({
    id,
    tooltip,
    defaultHtml: (d: any) =>
      `${d.data[y.key]} <br/> ${x
        .map((field) => `${field.key}: ${d.data[field.key]}`)
        .join('<br/>')}`,
  });
  /* eslint-disable */
  const formatMapping = {
    '%': '.0%',
    $: '($.2f',
    tri: ',.2r',
    hex: '#x',
    SI: '.2s',
  };

  const refreshChart = useCallback(() => {
    const svg = select(`#${id}`);
    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const g = svg.append('g');

    const yFn = scaleBand()
      .domain(data.map((d) => d[y.key]))
      .range([
        margin.top + padding.top,
        height - margin.bottom - padding.bottom,
      ])
      .padding(padding.bar);

    const xFnRange =
      direction === 'left'
        ? [width - margin.right - padding.right, margin.left + padding.left]
        : [margin.left + padding.left, width - margin.right - padding.right];

    const xFn = scaleLinear()
      // @ts-ignore
      .domain([0, max(data.map((d) => sum(x.map((value) => d[value.key]))))])
      .range(xFnRange);

    x.reverse();

    const dataStacked = stack().keys(x.map((column) => column.key))(data);

    transition();
    // @ts-ignore
    g.append('g')
      .selectAll('g')
      .data(dataStacked)
      .enter()
      .append('g')
      // @ts-ignore
      .attr('class', (d: any) => x[d.index].className)
      .selectAll('rect')
      .data((d: StackedArrayItem[], index: number) => {
        return d.map((v: StackedArrayItem) => {
          return {
            start: v[0],
            end: v[1],
            data: v.data,
            index,
          };
        });
      })
      .enter()
      .append('rect')
      .attr('x', () => xFn(0))
      .attr('y', (d: StackedDataItem) => {
        return (
          (yFn(d.data[y.key]) || 0) +
          (waterfall ? (yFn.bandwidth() / x.length) * d.index : 0)
        );
      })
      .attr('width', (d: StackedDataItem) =>
        drawing?.duration ? 0 : xFn(d.end) - xFn(d.start)
      )
      .attr('height', waterfall ? yFn.bandwidth() / x.length : yFn.bandwidth())
      .on('mouseenter', onMouseOver)
      .on('mousemove', onMouseMove)
      .on('mouseleave', onMouseLeave)
      .transition()
      .duration(drawing?.duration || 0)
      .delay((_, i) => (drawing?.delay || 0) * i)
      .attr('x', (d: StackedDataItem) => xFn(d.start))
      .attr('width', (d: StackedDataItem) => xFn(d.end) - xFn(d.start));

    const xAxis = x.some((column) => column.axis === 'top')
      ? // @ts-ignore
        axisTop(xFn).ticks(x.axisTicks || 5)
      : // @ts-ignore
        axisBottom(xFn).ticks(x.axisTicks || 5);

    tickFormat &&
      xAxis.tickFormat(
        // @ts-ignore
        formatMapping[tickFormat]
          ? // @ts-ignore
            format(formatMapping[tickFormat])
          : format(tickFormat)
      );

    const xAxisG = g.append('g').attr('class', 'axis--x axis ');

    const yAxis = direction === 'left' ? axisRight(yFn) : axisLeft(yFn);
    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${direction === 'left' ? width : margin.left},0)`
      );

    xAxisG
      .attr(
        'transform',
        // @ts-ignore
        `translate(0, ${x.axis === 'top' ? marginTop : height - margin.bottom})`
      )
      .call(xAxis);

    yAxisG.call(yAxis);
  }, [
    data,
    id,
    direction,
    padding,
    margin,
    referenceLines,
    waterfall,
    x,
    tickFormat,
    y,
    tooltip,
    drawing,
    dataLabel,
  ]);
  /* eslint-enable */

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll(`#tooltip-${id}`).remove();
    };
  }, [data, refreshChart]);
  return (
    <>
      <svg
        id={id}
        className={mergeTailwindClasses(defaultChartClassNames, className)}
      />
    </>
  );
};

export default BarChartStacked;
