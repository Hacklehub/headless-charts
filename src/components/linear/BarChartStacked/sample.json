[
  {
    "name": "Team A",
    "value": 10,
    "reading": 6
  },
  {
    "name": "Team B",
    "value": 20,
    "reading": 12
  },
  {
    "name": "Team C",
    "value": 15,
    "reading": 8
  },
  {
    "name": "Team D",
    "value": 18,
    "reading": 15
  }
]
