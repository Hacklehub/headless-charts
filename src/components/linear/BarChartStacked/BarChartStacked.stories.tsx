/* eslint-disable @typescript-eslint/no-explicit-any */
import BarChartStacked from '.';
import { Meta } from '@storybook/react';
import data from './sample.json';

export default {
  title: 'Linear/BarChartStacked',
  component: BarChartStacked,
  tags: ['autodocs'],
} as Meta;

export const Default = {
  args: {
    data,
    id: 'bar-chart-stacked-default',
    x: [
      {
        key: 'reading',
      },
      { key: 'value' },
    ],
    y: { key: 'name' },
  },
};

export const Styled = {
  args: {
    ...Default.args,
    id: 'bar-chart-stacked-styled',
    className: 'bg-gray-100 rounded',
    padding: {
      top: 20,
      right: 20,
      bottom: 20,
      left: 5,
      bar: 0.1,
    },
    margin: {
      top: 0,
      right: 40,
      bottom: 40,
      left: 60,
    },
    x: [
      {
        key: 'reading',
        className: 'fill-purple-700',
      },
      { key: 'value', className: 'fill-purple-400' },
    ],
    y: { key: 'name', className: 'text-red-500', padding: 10 },
  },
};

export const Tooltip = {
  args: {
    ...Styled.args,
    id: 'bar-chart-stacked-tooltip',
    tooltip: {
      className: 'bg-gray-800 text-white p-2 rounded',
      keys: ['name', 'reading', 'value'],
    },
  },
};

export const CustomTooltip = {
  args: {
    ...Styled.args,
    id: 'bar-chart-custom-tooltip',
    tooltip: {
      className: 'bg-gray-800 text-white p-2 rounded',
      html: (d: any) => {
        return `
            <div class="flex flex-col">
              <div class="text-lg text-center">${d.name}</div>
              <div class="text-sm text-center">${d.reading}</div>
              <div class="text-sm text-center">${d.value}</div>
            </div>
          `;
      },
    },
  },
};

export const WithDrawing = {
  args: {
    ...Styled.args,
    id: 'bar-chart-stacked-drawing',
    drawing: {
      duration: 1000,
    },
  },
};

export const WithDrawingDelay = {
  args: {
    ...Styled.args,
    id: 'bar-chart-stacked-drawing-delay',
    drawing: {
      duration: 1000,
      delay: 100,
    },
  },
};
