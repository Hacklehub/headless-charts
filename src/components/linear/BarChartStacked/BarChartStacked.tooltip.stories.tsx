import BarChartStacked from '.';
import { Meta } from '@storybook/react';
import data from './sample.json';

/**
 * Stacked bar charts are used to compare values across categories by using horizontal bars, with some grouping. Stacked bar charts additionally show proportions within categories.
 *
 *
 * */

const meta: Meta<typeof BarChartStacked> = {
  title: 'Linear/BarChartStacked/Tooltip',
  component: BarChartStacked,
  tags: ['autodocs'],
  args: {
    data,
  },
};

export default meta;

export const Tooltip = {
  args: {
    id: 'bar-chart-tooltip-default',
    x: [
      {
        key: 'macbook',
        className: 'fill-purple-800',
      },
      {
        key: 'iphone',
        className: 'fill-purple-600',
      },
      {
        key: 'ipad',
        className: 'fill-purple-400',
      },
      {
        key: 'wearables',
        className: 'fill-purple-300',
      },
      {
        key: 'services',
        className: 'fill-purple-200',
      },
    ],
    y: { key: 'year' },
    tooltip: {},
  },
};

export const TooltipStyled = {
  args: {
    ...Tooltip.args,
    id: 'bar-chart-tooltip-styled',
    tooltip: {
      className: 'bg-gray-100 rounded p-2',
    },
  },
};

export const TooltipCustomKeys = {
  args: {
    ...Tooltip.args,
    id: 'bar-chart-tooltip-custom-keys',
    tooltip: {
      className: 'bg-gray-100 rounded p-2',
      keys: ['year', 'macbook'],
    },
  },
};

export const TooltipCustomHtml = {
  args: {
    ...Tooltip.args,
    id: 'bar-chart-tooltip-custom-html',
    tooltip: {
      className: 'bg-gray-100 rounded p-2',
      html: (d: any) =>
        `<p className="text-xl">${d.data.year}</p><p classname="text-xs"> macbooks sold $${d.data.macbook}B</p>`,
    },
  },
};
