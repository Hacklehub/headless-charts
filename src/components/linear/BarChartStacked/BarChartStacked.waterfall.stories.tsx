import BarChartStacked from '.';
import { Meta } from '@storybook/react';
import data from './sample.json';

/**
 * Stacked bar charts are used to compare values across categories by using horizontal bars, with some grouping. Stacked bar charts additionally show proportions within categories.
 *
 *
 * */

const meta: Meta<typeof BarChartStacked> = {
  title: 'Linear/BarChartStacked/Waterfall',
  component: BarChartStacked,
  tags: ['autodocs'],
};

export default meta;

export const Waterfall = {
  args: {
    data,
    id: 'bar-chart-stacked-default',
    x: [
      {
        key: 'reading',
      },
      { key: 'value' },
    ],
    y: { key: 'name' },
    waterfall: true,
  },
};

export const Styled = {
  args: {
    ...Waterfall.args,
    id: 'bar-chart-stacked-styled',
    className: 'bg-gray-100 rounded',
    padding: {
      top: 20,
      right: 20,
      bottom: 20,
      left: 5,
      bar: 0.1,
    },
    margin: {
      top: 0,
      right: 40,
      bottom: 40,
      left: 60,
    },
    x: [
      {
        key: 'reading',
        className: 'fill-purple-700',
      },
      { key: 'value', className: 'fill-purple-400' },
    ],
    y: { key: 'name', className: 'text-red-500', padding: 10 },
  },
};

export const WithDrawing = {
  args: {
    ...Styled.args,
    id: 'waterfall-drawing-styled',
    drawing: {
      duration: 1000,
    },
  },
};

export const DrawingWithDelay = {
  args: {
    ...Styled.args,
    id: 'waterfall-drawing-delay',
    drawing: {
      duration: 1000,
      delay: 200,
    },
  },
};
