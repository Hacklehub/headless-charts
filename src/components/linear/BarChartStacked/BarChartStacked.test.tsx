import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import BarChartStacked from '.';
import data from './sample.json';

describe('BarChartStacked', () => {
  it('renders barchartstacked component for snapshot test', () => {
    render(
      <BarChartStacked
        data={data}
        id='barChartStacked-test'
        x={[
          {
            key: 'reading',
            className: '',
          },
          { key: 'value', className: '' },
        ]}
        y={{ key: 'name', className: '' }}
      />
    );
    const barChartStacked = screen.getByTestId('barChartStacked');
    expect(barChartStacked).toMatchSnapshot();
  });
  it('renders barchartstacked component', () => {
    render(
      <BarChartStacked
        data={data}
        id='barChartStacked-test'
        x={[
          {
            key: 'reading',
            className: '',
          },
          { key: 'value', className: '' },
        ]}
        y={{ key: 'name', className: '' }}
      />
    );
    const barChartStacked = screen.getByTestId('barChartStacked');
    expect(barChartStacked).toBeInTheDocument();
  });
});
