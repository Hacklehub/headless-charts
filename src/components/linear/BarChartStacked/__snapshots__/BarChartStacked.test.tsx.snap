// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarChartStacked > renders barchartstacked component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="barChartStacked"
  id="barChartStacked-test"
>
  <g>
    <g>
      <g
        class=""
      >
        <rect
          height="9.767441860465116"
          width="-24.242424242424242"
          x="60"
          y="6.04651162790698"
        />
        <rect
          height="9.767441860465116"
          width="-48.484848484848484"
          x="60"
          y="-7.906976744186046"
        />
        <rect
          height="9.767441860465116"
          width="-36.36363636363636"
          x="60"
          y="-21.86046511627907"
        />
        <rect
          height="9.767441860465116"
          width="-43.63636363636363"
          x="60"
          y="-35.81395348837209"
        />
      </g>
      <g
        class=""
      >
        <rect
          height="9.767441860465116"
          width="-14.545454545454547"
          x="60"
          y="6.04651162790698"
        />
        <rect
          height="9.767441860465116"
          width="-29.090909090909093"
          x="60"
          y="-7.906976744186046"
        />
        <rect
          height="9.767441860465116"
          width="-19.393939393939398"
          x="60"
          y="-21.86046511627907"
        />
        <rect
          height="9.767441860465116"
          width="-36.36363636363637"
          x="60"
          y="-35.81395348837209"
        />
      </g>
    </g>
    <g
      class="axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M60.5,6V0.5H-19.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(60.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(48.37878787878788,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(36.25757575757576,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(24.136363636363637,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          15
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(12.015151515151516,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-0.10606060606060552,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          25
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-12.227272727272723,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          30
        </text>
      </g>
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(60,0)"
    >
      <path
        class="domain"
        d="M-6,20.5H0.5V-39.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,10.930232558139538)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team A
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-3.0232558139534884)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-16.97674418604651)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team C
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-30.930232558139537)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Team D
        </text>
      </g>
    </g>
  </g>
</svg>
`;
